app/models/section.rb

```diff
@@ -3,20 +3,60 @@ class Section < ActiveRecord::Base
   # Associations
   belongs_to :phase
   belongs_to :organisation
-  has_many :questions, :dependent => :destroy
+  has_many :questions, dependent: :destroy
   has_one :template, through: :phase
 
-  #Link the data
-  accepts_nested_attributes_for :questions, :reject_if => lambda {|a| a[:text].blank? },  :allow_destroy => true
+  # Link the data
+  accepts_nested_attributes_for :questions, reject_if: lambda { |a| a[:text].blank? },  allow_destroy: true
 
   attr_accessible :phase_id, :description, :number, :title, :published,
                   :questions_attributes, :organisation, :phase, :modifiable,
-                  :as => [:default, :admin]
+                  as: [:default, :admin]
 
-  validates :phase, :title, :number, presence: {message: _("can't be blank")}
+  ##
+  # Validations
+  validates :phase, :title, presence: { message: _("can't be blank") }
+
+  validates :number, presence: { message: _("can't be blank") },
+                     uniqueness: { scope: :phase_id }
 
   before_validation :set_defaults
 
+  ##
+  # Scopes
+
+  # The sections for this Phase that have been added by the admin
+  #
+  # @!scope class
+  # @return [ActiveRecord::Relation] Returns the sections that are modifiable
+  scope :modifiable, -> { where(modifiable: true) }
+
+  # The sections for this Phase that were part of the original Template
+  #
+  # @!scope class
+  # @return [ActiveRecord::Relation] Returns the sections that aren't modifiable
+  scope :not_modifiable, -> { where(modifiable: false) }
+
+  # =================
+  # = Class methods =
+  # =================
+
+  def self.update_numbers!(*ids, phase:)
+    # Ensure only section ids belonging to this Phase are included.
+    ids    = ids.map(&:to_i) & phase.section_ids
+    return if ids.empty?
+    # Build an Array with each ID and its relative position in the Array
+    values = ids.each_with_index.map { |id, i| "(#{id}, #{i + 1})" }.join(", ")
+    # Run a single UPDATE query for all records.
+    query  = <<~SQL
+      UPDATE #{table_name} \
```

Comment:
does table_name get infered from the model here?

Reply:
Yes. Table name is defined as part of ActiveRecord. This is preferred, since you might change the table name in future.

Reply:
I like the single query update. Much more efficient.

Reply:
Just tried running this on roadmap-stg and it looks like MySQL doesn't like the structure of the statement:
```bash
ActiveRecord::StatementInvalid: Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM (VALUES (3599, 1), (3600, 2), (3601, 3), (3602, 4), (3603, 5), (3604, 6), (' at line 1: UPDATE sections         SET number = svals.number         FROM (VALUES (3599, 1), (3600, 2), (3601, 3), (3602, 4), (3603, 5), (3604, 6), (3605, 7)) AS svals(id, number)         WHERE svals.id = sections.id;
```